### ISSUE TITLE
Yolo11 OBB: How to interpret and postprocess the output predictions

### ISSUE THREAD

**Issue by Berni11** [ultralytics/ultralytics#19088]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have trained a Yolo11n-Obb model on a custom dataset and exported it into the tflite format. I have tested the tflite model with the Ultralytics library and it works as expected.
I want to use it in an Flutter Android app with Tensorflow Lite, which means I don't have access to Ultralytics Python library and it's postprocessing.

After inference I get a shape of the size (1, 10, 8400). I have transposed it into the shape (8400, 10) which gives me the values for each predictions. If I understand it correctly the values are: cx, cy, width, height, angle, and confidence score for each class.

My question is: what exactly do I have to do in the postprocessing with the predicted values? Do I apply a sigmoid function or something similar to the values?

I wasn't able to find anything in the Documentation.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Berni11, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com/) where you'll find resources on [Python](https://docs.ultralytics.com/usage/python/), [CLI](https://docs.ultralytics.com/usage/cli/) usage, and much more. These may provide helpful insights for your question.

As you are working with a YOLO11n-OBB model in TensorFlow Lite for a Flutter Android app, there might be nuances to adapting the postprocessing function outside the Ultralytics Python ecosystem. Currently, we do not have native documentation for implementing postprocessing in your scenario.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to assist us in debugging.

For your ❓ Question, providing more detailed information such as the raw outputs from the tflite model inference and how you're processing those outputs might help clarify your use case. Ensure you're following [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to validate your setup.

Join the Ultralytics community to explore your question further:
- 🌟 [Discord](https://discord.com/invite/ultralytics) for real-time discussions
- 💬 [Discourse](https://community.ultralytics.com/) for in-depth, community-driven conversations
- 👾 [Subreddit](https://reddit.com/r/Ultralytics) to connect with other users and share knowledge 

## Upgrade

Ensure you're working with the latest version of `ultralytics` along with corresponding dependencies in a [Python>=3.8](https://www.python.org/) environment and [PyTorch>=1.8](https://pytorch.org/get-started/locally/). Upgrading can resolve potential issues as well as provide access to the newest features:

```bash
pip install -U ultralytics
```

## Environments

YOLO runs seamlessly in various environments:
- **Notebooks** with GPU support: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: Refer to [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **AWS** Deep Learning AMI: Refer to [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: Accessible via [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a> 

If this badge is green, all [Ultralytics CI tests](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) are currently passing. CI runs daily to ensure robust operation across platforms and modes.

This is an automated response 😊. An Ultralytics engineer will review your issue and provide additional assistance soon.

**Comment by Y-T-G** [comment]
You can look at the post-processing in examples.

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-TFLite-Python

OBB is complicated because you need to perform rotated NMS. You can try exporting with `nms=True`. That would generate an output which doesn't require NMS. So you would just need to apply confidence threshold.

The output format with `nms=True` would be `x1, y1, x2, y2, conf, label`. So it's pretty easy to process. The output coordinates are normalized. So you would need to multiply with the x1 and x2 with input width and y1 and y2 with input height to denormalize it. The example above shows that.

**Comment by Berni11** [comment]
Thank you for the reply. I've tried it now with the nms export enabled and it seems to work. But my output shape is (1, 300, 7). What would be the 7th value? Is it the angle because of the obb model?

**Comment by Y-T-G** [comment]
Yes. It's angle in radians.

**Comment by Berni11** [comment]
After testing I have noticed that most detections have an angle of 0. Is this normal?

How do I filter out unwanted/bad detections? I have tried filtering by score but there are still overlaping detections.

**Comment by Y-T-G** [comment]
Only the ones above the  confidence threshold will have valid angles. Not all the 300 detections are valid.

**Comment by Berni11** [comment]
I have the problem that I get multiple detections for the same object, but the detection which has an angle has a lower score than the other detections.
Do I need to apply a sigmoid function to the scores?
Also the angles are always tilted too much. Any way to improve this?

**Comment by Berni11** [comment]
I wasn't able to get the nms enabled model running properly in Ultrylytics to test it.

The nms enabled model would work but has the angle problem. Inference is done with an image with the shape (N,H,W,C), like in the python example.

There is probably something not working properly but I don't know what it could be. For example the normal model (without nms enabled) in Ultralytics, the detected objects have scores of around 0.50 to 0.80, but in my Flutter app the highest score of the same class is 0.012108506634831429 with no filtering applied yet.

**Comment by Berni11** [comment]
After a bit of testing and searching I've found the same detections from the Ultralytics library in my output tensor. The reason why I couldn't find them was the the classes had a different order.
Instead of the order: class0, class1, class2, class3, class4
It was: class1, class2, class3, class4, class0

Now the only problem I have is the angle, as I don't know what kind of normalization it has.

Here are for comparison the Ultralytics detections and my detections (x, y, width, height, angle):
Image 1 Ultralytics:
[532.0693, 474.7077,  48.1370,  23.3950,   2.9284]
[259.8898, 460.6205,  65.9054,  29.8028,   0.9929]
[174.2333, 629.0838, 105.7762,  40.0888,   2.8934]

Image 1 (mine):
532.0694, 474.7077, 48.137, 23.3951, 7.61020373829524e-7
259.8898, 460.6205, 65.9054, 29.8029, 3.814522813172516e-7
174.2334, 629.0837, 105.7763, 40.089, 0.0000012063470649081864

Image 2 Ultralytics:
[299.0104, 117.0404,  54.6779,  24.8559,   1.9679]
[217.7603,  66.8476,  12.2765,  11.4729,   1.5702]

Image 2 (mine)(for some reason width and height are flipped):
299.0104, 117.0405, 24.856, 54.678, 2.7523708467924735e-7
217.7603, 66.8476, 11.4729, 12.2765, 4.962844855072035e-7

**Comment by Y-T-G** [comment]
Are you talking about the model with NMS or without.

The angles aren't normalized.

**Comment by Y-T-G** [comment]
The angle is the last value if you're using the exported model without NMS.

`x, y, w, h, class0, class1, ..., class9, angle`

**Comment by Berni11** [comment]
The model without NMS.

I thought it was: x, y, w, h, angle, class0, ....

I will test and report.

**Comment by Berni11** [comment]
The updated values I get now look much better:

Image 1
532.0694, 474.7077, 48.137, 23.3951, -0.21314267814159393
259.8898, 460.6205, 65.9054, 29.8029, 0.9928745627403259
174.2334, 629.0837, 105.7763, 40.089, -0.24818480014801025

Image 2
299.0104, 117.0405, 24.856, 54.678, 0.3971405625343323
217.7603, 66.8476, 11.4729, 12.2765, -0.0005990266799926758

I see that some have pi or pi/2 added to them in the Ultralytics output. is this a necessary step?

**Comment by Y-T-G** [comment]
It's conversion from radians to degrees

**Comment by Berni11** [comment]
ok. thank you for the help

**Comment by glenn-jocher** [comment]
You're welcome! For OBB angles in the TFLite output, they're provided directly in radians (no sigmoid needed) as shown in our [exporter code](https://docs.ultralytics.com/reference/engine/exporter/#ultralytics.engine.exporter.NMSModel.forward). If angles appear off, ensure you're using the correct confidence threshold and consider verifying your training data rotations. The Ultralytics team is here if you have further questions! 🚀

**Comment by ktobah** [comment]
@Berni11 Could you please share your preprocessing/postprocessing codes?
